contracts/System.sol

```diff
@@ -80,4 +122,171 @@ contract System {
+
+  function _lightClient() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.lightClient);
+    } else {
+      return _LIGHT_CLIENT_ADDR;   
+    }
+  }
+
+  function _relayerHub() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.relayerHub);
+    } else {
+      return _RELAYER_HUB_ADDR;   
+    }
+  }
+
+  function _candidateHub() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.candidateHub);
+    } else {
+      return _CANDIDATE_HUB_ADDR;   
+    }
+  }
+
+  function _govHub() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.govHub);
+    } else {
+      return _GOV_HUB_ADDR;   
+    }
+  }
+
+  function _pledgeAgent() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.pledgeAgent);
+    } else {
+      return _PLEDGE_AGENT_ADDR;   
+    }
+  }
+
+  function _burn() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.burn);
+    } else {
+      return _BURN_ADDR;   
+    }
+  }
+
+  function _foundation() view internal returns (address) {
+    if (_isLocalTestNode()) {
+      return _notNull(_ext().addrs.foundation);
+    } else {
+      return _FOUNDATION_ADDR;   
+    }
+  }
+
+  function updateContractAddr(address validatorSet_, address slash_, address systemReward_, 
+                              address lightClient_, address relayerHub_, address candidateHub_, 
+                              address govHub_, address pledgeAgent_, address burn_, address foundation_) 
+           external onlyLocalTestMode onlyTestModeDeployer updateContractAddrCalledOnce {
+
+    require(!_testModeAddressesWereSet(), "test-mode addresses already set");
+
+    assert( validatorSet_ != address(0));
+    assert( slash_ != address(0));
+    assert( systemReward_ != address(0));
+    assert( lightClient_ != address(0));
+    assert( relayerHub_ != address(0));
+    assert( candidateHub_ != address(0));
+    assert( govHub_ != address(0));
+    assert( pledgeAgent_ != address(0));
+    assert( burn_ != address(0));
+    assert( foundation_ != address(0));
+
+    _ext().addrs = Addresses({
+        validatorSet: validatorSet_,
+        slash: slash_,
+        systemReward: systemReward_,
+        lightClient: lightClient_,
+        relayerHub: relayerHub_,
+        candidateHub: candidateHub_,
+        govHub: govHub_,
+        pledgeAgent: pledgeAgent_,
+        burn: burn_,
+        foundation: foundation_
+    });
+  }
+
+  function _notNull(address addr) private pure returns (address) {
+    require(addr != address(0), "address is null");
+    return addr;
+  }
+
+  function _isLocalTestNode() private view returns (bool) {
+    return block.chainid != CORE_MAINNET && block.chainid != CORE_TESTNET; // any network which is neither Core mainnet or testnet
```

Comment:
pls double-check me here: is there any chance that real assets will be placed in a network with chainid other than 1116?

Reply:
I could be missing your point here but for this to happen shouldn't the contracts be deployed on the CoreL2 network, which brings us to a state similar to the codebase-clone from your other comment?